File: apps/push-service/src/main/java/com/example/chat/push/domain/PushMessageRepository.java
package com.example.chat.push.domain;

import jakarta.persistence.LockModeType;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;

import java.time.LocalDateTime;
import java.util.List;

public interface PushMessageRepository extends JpaRepository<PushMessage, Long> {

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query("SELECT p FROM PushMessage p WHERE p.status = 'PENDING' AND p.createdAt < :cutoff ORDER BY p.createdAt ASC")
    List<PushMessage> findPendingForProcessing(LocalDateTime cutoff);
}

File: apps/push-service/src/main/kotlin/com/example/chat/push/domain/PushMessageRepository.kt
package com.example.chat.push.domain

import jakarta.persistence.LockModeType
import java.time.LocalDateTime
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.data.jpa.repository.Query

interface PushMessageRepository : JpaRepository<PushMessage, Long> {
    @Lock(LockModeType.PESSIMISTIC_WRITE)
    @Query(
            "SELECT p FROM PushMessage p WHERE p.status = 'PENDING' AND p.createdAt < :cutoff ORDER BY p.createdAt ASC"
    )
    fun findPendingForProcessing(cutoff: LocalDateTime): List<PushMessage>
}
